ddtrace/internal/datadog/profiling/dd_wrapper/dd_wrapper.md

```diff
@@ -0,0 +1,74 @@
+
+Although this sampler operates in its own thread, since its fundamental interaction requires taking the GIL, it never samples at a time when other threads are on-CPU (except if they're off-GIL in C code or in io or whatever, but that isn't an issue).
+
+### Lock sampler
+
+This patches some common lock implementations to perform sampling.  As such, the sampling occurs from the active application thread.
+
+### Memory sampler
+
+This is very similar to the lock profiler, except for allocations.
+
+
+## Current state
+
+For various reasons, we're validating profiling concepts where as much (or all!) of the work is done from native threads.  This is a bit of a departure from the traditional dd-trace-py model, but the technique is well-represented in the community.  This is comprised of two main parts right now
+
+* Moving the stack sampler to a native, off-GIL thread
+* Moving the collection apparatus to native code
+
+The problem here is that in Python it's fairly common to periodically call `fork()` as a worker lifecycle management technique.  When this happens, then in the child:
+
+* Execution resumes from the same point as the parent.  One way to think about this is, execution resumes from the thread that called fork
+* The heap is configured to be copied-on-write.
+
+This means any intermediate heap-stored state managed by a sibling thread (i.e., a thread which did not call `fork()`) may become corrupt in the child.  We need to protect or re-initialize such state when possible.
+
+## Data organization in this code
+
+Roughly, the data is organized in the following way
+
+### Samples
+
+Samples are stored on the heap.  They are managed by a GlobalCache object.  GlobalCache provides a `get()` method which takes a thread ID and returns a Sample (or creates a sample if needed).  Samples form the basis of a single collection operation.
+
+Samples are interacted with transactionally, for example
+
+1. `ddup_start_sample()` (every sample begins with a call to `start_sample()`)
+2. adding frames, walltime, tags, etc
+3. ddup_flush_sample() (every sample ends with a call to `flush_sample()`)
+
+The act of flushing a sample stores its data in a ddog_prof_Profile object (which is wrapped by Profile in this code).
+
+There's one wrinkle here.  The navigation through frame data (unwinding) may result in temporary strings.  We need to cache these strings.  In order to minimize overhead, strings are cached (and de-duplicated) in a cache attached to the Profile rather than the Sample.
+
+After `start_sample()`, the current sample in the current thread is saved in `thread_local` storage, removing the need for the thread to call into global cache.
```

Comment:
This is a fundamental question and I answered it by dropping the cache entirely.

Here's the deal, as far as I understand it.

On Linux, thread-local storage is just another form of heap allocation.  The threading runtime (NPTL) sets it up when a thread is made and associates that storage to a given thread at runtime.  C++ hijacks this mechanism for `thread_local`.  Post-fork(), nothing special is done to these arenas--it's heap memory, so the user could have sent pointers around in their application, it can't get cleaned up.

On the other hand, provided the user did not do a silly thing (like _not_ propagate pointers to TLS memory), those pages (I believe TLS areas are page-aligned and allocated individually) are CoW, meaning that the marginal overhead to the system for keeping them in the child is nearly zero.

I knew this in my first cut and struggled to form a conviction, but after reading your comment I realized that in the average case a user calling `fork()` will be running a fork-based worker pool with regeneration.  The children won't spawn children of their own.  I think the marginal value of preventing useless TLS copies is very close to zero.  Ignoring this problem simplifies the code a lot.